pkg/gcp/client/storage.go

```diff
@@ -56,18 +57,54 @@ func NewStorageClientFromSecretRef(ctx context.Context, c client.Client, secretR
 	return NewStorageClient(ctx, serviceAccount)
 }
 
-func (s *storageClient) CreateBucketIfNotExists(ctx context.Context, bucketName, region string) error {
-	if err := s.client.Bucket(bucketName).Create(ctx, s.serviceAccount.ProjectID, &storage.BucketAttrs{
-		Name:     bucketName,
-		Location: region,
+// CreateOrGetAndUpdateBucket creates a new GCS bucket if it does not already exist.
+// If the bucket already exists and the provided immutability settings differ from the current
+// retention policy, the retention policy is updated accordingly.
+//
+// Parameters:
+//   - ctx: The context for the request.
+//   - bucketName: The name of the bucket to create or update.
+//   - region: The region where the bucket will be created.
+//   - imSettings: The immutability settings to apply to the bucket. If nil, no retention policy is set.
+//
+// Returns:
+//   - error: An error if the bucket creation or update fails, or nil if successful.
+func (s *storageClient) CreateOrGetAndUpdateBucket(ctx context.Context, bucketName, region string, config *apisgcp.BackupBucketConfig) error {
+	var retentionPolicy *storage.RetentionPolicy
+	if config != nil {
+		retentionPolicy = &storage.RetentionPolicy{
+			RetentionPeriod: config.Immutability.RetentionPeriod.Duration,
+			IsLocked:        true,
+		}
+	}
+
+	err := s.client.Bucket(bucketName).Create(ctx, s.serviceAccount.ProjectID, &storage.BucketAttrs{
+		Name:            bucketName,
+		Location:        region,
+		RetentionPolicy: retentionPolicy,
 		UniformBucketLevelAccess: storage.UniformBucketLevelAccess{
 			Enabled: true,
 		},
 		SoftDeletePolicy: &storage.SoftDeletePolicy{
 			RetentionDuration: 0,
 		},
-	}); err != nil {
+	})
+	if err != nil {
 		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == errCodeBucketAlreadyOwnedByYou {
+			// Bucket already exists, update retention policy if necessary
+			bucket := s.client.Bucket(bucketName)
+			attrs, err := bucket.Attrs(ctx)
+			if err != nil {
+				return err
+			}
+			if config != nil && config.Immutability != (apisgcp.ImmutableConfig{}) && (attrs.RetentionPolicy == nil || attrs.RetentionPolicy.RetentionPeriod != config.Immutability.RetentionPeriod.Duration) {
+				_, err = bucket.Update(ctx, storage.BucketAttrsToUpdate{
```

Comment:
If I am not missing anything, with the current implementation you will always be doing 2 API calls and I am curious if instead a "Get and Create or Update-if-necessary) is better. 

I am not opposed to keep the current implementation but if you do so, can you check if subsequent reconciliations after the update may cause issues ?
In some cases the clients may return an error if an update call "changed nothing" e.g. https://github.com/gardener/gardener-extension-provider-gcp/blob/d12986b67bdc2b70022aa6eb350a9eb431b35d6d/pkg/gcp/client/compute.go#L458

Reply:
> We need to get the attributes (metadata) of the bucket to determine whether the bucket already exists

Just checking with a GET call the GCP API should also suffice.

Reply:
Actually GetBucket also returns the `BucketAttrs` only.
`GetBucket(ctx context.Context, bucket string, conds *BucketConditions, opts ...storageOption) (*BucketAttrs, error)`
https://github.com/googleapis/google-cloud-go/blob/04738d8363b138d18b037cd761b2777acb311a2d/storage/client.go

Reply:
I've made changes as part of commit 75bc2f4b8ccf3a5d5764be15b7e03f2b71865a25. Now, the number of API calls made during different operations are as follows:
- During bucket creation, there are three API calls: GetBucketAttributes, Create, and Lock (if immutability is enabled).
- During bucket update, there are up to three API calls: GetBucketAttributes, Update, and Lock (if immutability is introduced as part of this update).
- During reconciliation, there is one API call: GetBucketAttributes.